From decf421412f95b7cdd023fb986799dd9f537d40a Mon Sep 17 00:00:00 2001
From: Hariprasad Kelam <hkelam@marvell.com>
Date: Thu, 25 Feb 2021 22:55:48 +0530
Subject: [PATCH 1277/1921] octeontx2-pf: fix vf ethtool sset count

Ethtool gets stats info from sset_count and get_strings callbacks.
Current code does not consider complete stats while calculating
sset count for VF devices.This patch fixes the same.

Fixes: 1f5499b99("octeontx2-pf: Backport cosmetic upstream changes")
Change-Id: Ib417891b5ca2fe1cfbd48d74ca8fe2da967242cf
Signed-off-by: Hariprasad Kelam <hkelam@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/46820
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
index 583071bcf6d9..d19ea599c635 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
@@ -1643,12 +1643,15 @@ static void otx2vf_get_ethtool_stats(struct net_device *netdev,
 static int otx2vf_get_sset_count(struct net_device *netdev, int sset)
 {
 	struct otx2_nic *vf = netdev_priv(netdev);
+	int qstats_count;
 
 	if (sset != ETH_SS_STATS)
 		return -EINVAL;
 
-	return otx2_n_dev_stats +
-	       otx2_n_queue_stats * (vf->hw.rx_queues + vf->hw.tot_tx_queues);
+	qstats_count = otx2_n_queue_stats *
+		       (vf->hw.rx_queues + vf->hw.tx_queues);
+
+	return otx2_n_dev_stats + otx2_n_drv_stats + qstats_count + 1;
 }
 
 static int otx2vf_get_link_ksettings(struct net_device *netdev,
-- 
2.31.1

